Accessibility: Improve HelpModal markup

tskarhed opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Removes the table in the HelpModal to make it more accessible for screen readers. With the help of the `Text` component, the contrast issue for the light theme is addressed. Also made it readable for extra small devices, because why not?

Before:
![Screenshot 2024-02-21 at 14 29 13](https://github.com/grafana/grafana/assets/1438972/5ed3cd37-9f1d-46a5-bd1e-4ed3ac9ba04a)
After:
![image](https://github.com/grafana/grafana/assets/1438972/cb7a4bb7-d58a-4bc3-a047-bcd231a308e1)

**Why do we need this feature?**

It is very heard to read for screenreaders.
 
**Who is this feature for?**

Screen reader users

**Which issue(s) does this PR fix?**:

Fixes #79516 

**Special notes for your reviewer:**

I used a definition list and tried it with the screen reader, but I am not convinced. It did work well with the CSS grid though.

There was a discussion about separation the keys with modifiers, but I don't think we should do it, as it highlights that they should be pressed at the same time as opposed to in succession. 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

tskarhed committed:
HelpModal: Make more accessible

tskarhed commented:
Should we write/design something that makes it easier to understand if you should press the keys at the same time or in order?

tskarhed commented:
/deploy-to-hg

tskarhed committed:
Remove console.log

ckbedwell reviewed (commented):
To be honest, I'm not that precious about this but I'd probably say I'd rather have a table over the Definition List. Two main reasons why:

1. Tables actually work really well with screen readers - in particular that you can easily navigate up and down columns which I think would be pretty desirable when as a user you want to see what shortcuts are available, _then_ look at what the shortcut is if you want to use it (for this reason I'd also think about changing the design and put the command first then the keys second  -- again minor and not sure if design will sign off on that change or not but all the other sites I've spot-checked do it this way.)

I should have made this clearer in the filed issue but the current implementation's main problem is the way the table header is implemented. Neither columns have their own header, just a shared one that spans two columns, so it can be disorientating for screen reader users when entering the table.

2. I think the screen reader implementation for definition lists is pretty annoying. There is clearly 8 commands which you have marked up correctly but the screen reader reads it as "16 items", as it considers each definition term and definition description as an item and doesn't think they are related 🤷 You have to go through all of the commands and keys associated before finding out what the next command is.

> There was a discussion about separation the keys with modifiers, but I don't think we should do it, as it highlights that they should be pressed at the same time as opposed to in succession.

I never noticed that they could be hit one after the other until this comment, I wasn't sure what you meant until I tested it 😅  So on this note, the problem with them listed like this is they are treated separately with VoiceOver (when I tested with NVDA they are read together, though, so it is platform specific -- might also be because of the settings I use for my screen readers).

You could go the route of rendering a grouped text version that is hidden visually (e.g. `className="sr-only"`) and putting `aria-hidden` on the visible keys so they aren't read out twice.

Lastly -- for screen reader users `Esc` is read out as `e - s - c` rather than escape, (same with `c - m - d`). The few keys that are capitals also aren't clear to screen reader users. It might be worth specifically mentioning it is `d shift + c` as an example.

tskarhed committed:
Handle custom keys for screen reader

tskarhed commented:
I added specific characters for command and shift. I am doing sr-only for ctrl and esc. Writing the whole word seems a bit excessive, especially since there is a convention around it.

tskarhed committed:
Rewrite using tables

tskarhed committed:
Increase gap

tskarhed commented on public/app/core/components/help/HelpModal.tsx:
```diff
@@ -140,86 +143,74 @@ export interface HelpModalProps {
+                {category}
+              </Text>
+            </div>
+            <dl className={styles.keysAndDescriptions}>
```
Using a `caption` for the table

tskarhed committed:
Merge branch 'main' into 79516-help-modal-a11y

tskarhed commented:
/deploy-to-hg

amy-super commented:
One little trick that might help the layout look a bit better would be to reorder the sections. So `Global` on the top left, `Dashboard` on the bottom left, `Time Range` on the top right, `Focused Panel` on the bottom right. 

I'd also change them to all to sentence case (`Focused panel`, `Time range`) to be a more consistent with Grafana styling.

tskarhed committed:
Merge branch 'main' into 79516-help-modal-a11y

tskarhed committed:
Change order of help categories

ckbedwell reviewed (commented):
Left one minor comment about the srOnly implementation to make it meet the official standard.

I also noticed that when I am zoomed in a bit I can't scroll up and down the content using just my keyboard. @eledobleefe [looked at this problem recently](https://github.com/grafana/design-system/pull/117/files#diff-596152d30a8b7f508b5c95c7eb146e27b695e362dbb4eda91745acbf30b17ab7R50-R54) and we agreed the best solution is to get modal authors to add `tabIndex={0}` to content containers (in this case the Grid component in the HelpModal) with no focusable elements.

One last thing I want to highlight so that if this ever gets flagged in a future audit we can say we are aware of it and have chosen to ignore this best practice (or maybe add it to be thought about in the future for these kind of designs? cc @amy-super).

[So it is best practice that when using a table, that the table headers are visible](https://dequeuniversity.com/rules/axe/4.1/empty-table-header). (Quick side node --  the way our `.sr-only` class is written Deque's automated dev tools think it is visible...). As previously discussed, this design could be implemented as a definition list, or just without table markup at all but we've chosen not to do that because we believe the experience is worse for users that way.

tskarhed committed:
HelpModal: Add tabIndex and imrpove sr-only display

tskarhed commented:
Okay, I've done this:
- Added tabIndex for scrolling with keyboard
- Changed the way ctrl and esc are read to screen readers

From my understanding, you are okay with the table headers not showing.

ckbedwell reviewed (approved):
LGTM! 🌟
